README.md
# kz-iin-validator

[![MIT License](https://img.shields.io/pypi/l/kz-iin-validator.svg?style=flat-square)](https://opensource.org/licenses/MIT)
[![PyPI version](https://img.shields.io/pypi/v/kz-iin-validator.svg?style=flat-square)](https://pypi.org/project/kz-iin-validator/)
[![PyPI downloads](https://img.shields.io/pypi/dm/kz-iin-validator.svg?style=flat-square)](https://pypi.org/project/kz-iin-validator/)
[![codecov](https://img.shields.io/codecov/c/github/ZhymabekRoman/kz-iin-validator?style=flat-square)](https://app.codecov.io/github/ZhymabekRoman/kz-iin-validator)
[![tests](https://img.shields.io/github/actions/workflow/status/ZhymabekRoman/kz-iin-validator/pytest.yml?branch=main&style=flat-square)](https://github.com/ZhymabekRoman/kz-iin-validator/actions)

**kz-iin-validator** - Библиотека для проверки и извлечения данных из ИИН на Python 3. ИИН - это Индивидуальный Идентификационный Номер граждан Республики Казахстан (РК) и юридических лиц РК

## Возможности:
- Production-ready библиотека
- 100% code coverage - Библиотека полностью покрыта тестами
- Zero-depency - Не требует дополнительных зависимостей
- Базовая валидация данных ИИН, таких как дата, месяц, год
- Генератор ИИН
- Проверка целостности по хэш сумме ИИН. Алгоритм формирования ИИН регулируется Постановлением правительства РК "Об утверждении Программы перехода на единый номер физического (юридического) лица (индивидуальный идентификационный номер (бизнес-идентификационный номер) в целях создания Национальных реестров идентификационных номеров Республики Казахстан" от 11 июня 2003 года N 565 - [Ссылка на источник приказа](https://adilet.zan.kz/rus/docs/P030000565_)

## Требования:
- Python 3.7 и выше * **
> \* судя по информации утилиты Vermin

> ** я могу адаптировать (забэкпортить) библиотеку для более старых версии Python, но стоит ли оно? Если что дайте мне знать

## TODO:
- Реализовать поддержку валидации БИН
- Улучшить документация
- Static type hint using Mypy

## Схема формирования ИИН:
<details>
  <summary>Схема</summary>
    <img src="https://raw.githubusercontent.com/ZhymabekRoman/kz-iin-validator/main/images/iin_schema.webp" alt="Schema IIN" />
</details>

## Установка:
### PyPI:
```bash
python3 -m pip install kz-iin-validator
```
### Из GIT репозиторий:
```bash
python3 -m pip install git+https://github.com/ZhymabekRoman/kz-iin-validator
```

## Использование:
### Валидация ИИН:
```python
from kz_iin_validator import safe_validate_iin, validate_iin

iin = "061211600012"
# use kz_iin_validator as in Golang (preferred in production)
validated_iin, error = safe_validate_iin(iin)
# or directly with exceptions raising:
# validated_iin = validate_iin(iin)

print(f"ИИН: {iin}")

# Print IIN owner gender:
print(f"Пол: {validated_iin.gender.name.upper()}")

# Print IIN owner born information:
print(f"День: {validated_iin.born_date.day}")
print(f"Месяц: {validated_iin.born_date.month}")
print(f"Год: {validated_iin.born_date.year}")

# or directly access to datetime object:
# validated_iin.born_date.datetime
```

### Генератор ИИН:
```python
from kz_iin_validator import generate_iin

generated_iin = generate_iin()
print(generated_iin) # return generated valid IIN as string
```

### Извлечь ИИН из текста:
```python
from kz_iin_validator import extract_iin

text = "Lorem ipsum dolor sit amet, consectetur adipiscing elit - 251112401047. Aliquam vel diam ac enim consequat rhoncus 700911204362!."

extracted_iin_list = extract_iin(text)
print(extracted_iin_list)
```


CHANGES.md
CHANGES
===============

kz-iin-validator 0.7.0
-----------

### Changes between 0.6.0 and 0.7.0
 * `weak_fast_check` argument from functions `validate_iin` and `safe_validate_iin` was removed
 * IIN extractor from text was implemented in function `extract_iin`.

kz-iin-validator 0.6.0
-----------

### Changes between 0.5.0 and 0.6.0
 * Function `validate_iin` was renamed to `safe_validate_iin`
 * Private function `_validate_iin` was renamed to `validate_iin`
 * Private property `_datetime` was renamed to `datetime`
 * Argument `raise_exception=False` from `safe_validate_iin` was deleted
 * Fix typos: `lenght` to `length`
 * Minor fixes


tests/test_basic.py
import random

import pytest

from kz_iin_validator import IINValidateError, generate_iin, safe_validate_iin, validate_iin, extract_iin
from kz_iin_validator.utils import is_digit_string


def test_valid_iin():
    for _ in range(20000):
        iin = generate_iin()
        validated_iin = validate_iin(iin)
        assert validated_iin is not None

    for _ in range(20000):
        iin = generate_iin()
        validated_iin = safe_validate_iin(iin)
        assert validated_iin is not None


def test_non_valid_iin():
    non_valid_iin_list = ["0000000", 868795643215, "sdf867", "ajd85gt6h5yj", "/*-/*-gd098j6435", "821214240092"]

    _iin = list(generate_iin())
    _iin[6] = str(random.randint(7, 9))
    non_correct_centry_iin = "".join(_iin)
    non_valid_iin_list.append(non_correct_centry_iin)

    _iin = list(generate_iin())
    _iin[11] = str(int(_iin[11]) - 1)
    non_correct_control_sum_iin = "".join(_iin)
    non_valid_iin_list.append(non_correct_control_sum_iin)

    _iin = list(generate_iin())
    _iin[2:4] = str(random.randint(13, 99))
    non_correct_month_iin = "".join(_iin)
    non_valid_iin_list.append(non_correct_month_iin)

    _iin = list(generate_iin())
    _iin[4:6] = str(random.randint(32, 99))
    non_correct_date_iin = "".join(_iin)
    non_valid_iin_list.append(non_correct_date_iin)

    _iin = list(generate_iin())
    _iin[6] = str(0)
    non_correct_gender_iin = "".join(_iin)
    non_valid_iin_list.append(non_correct_gender_iin)

    for iin in non_valid_iin_list:
        with pytest.raises(IINValidateError):
            validate_iin(iin)

        iin_result, err = safe_validate_iin(iin)
        assert err is not None

        iin_result, err = safe_validate_iin(iin)
        assert err is not None


def test_utils():
    assert is_digit_string("1316456513")
    assert is_digit_string("sdfkjdsf") is False
    assert is_digit_string("656543413")
    assert is_digit_string("grarthaeg") is False


kz_iin_validator/iin_generator.py
import random


def generate_iin():
    """
    Generates a random Individual Identification Number (IIN).

    Returns:
        str: The randomly generated IIN.

    """
    # year = random.randint(1800, 2099)
    year = random.randint(0, 99)
    month = random.randint(1, 12)

    # TODO: manually generate gender instead generating with centry
    # gender = random.choice(list(IINGender))

    centry_and_gender_list = {1: 18, 2: 18, 3: 19, 4: 19, 5: 20, 6: 20}
    centry_and_gender_code = random.choice(list(centry_and_gender_list))
    centry_code = centry_and_gender_list[centry_and_gender_code]

    month_dict = {4: 30, 6: 30, 9: 30, 11: 30, 2: 28}
    day_bound = month_dict.get(month, 31)

    year_centry = int(f"{centry_code}{year:02}")

    if day_bound == 28 and ((year_centry % 4 == 0 and year_centry % 100 != 0) or year_centry % 400 == 0):
        day_bound = 29

    day = random.randint(1, day_bound)

    # zfill
    year_str = f'{year:02d}'
    month_str = f'{month:02d}'
    day_str = f'{day:02d}'

    dob = f"{year_str}{month_str}{day_str}"

    individual_number = f"{random.randint(1, 9999):04d}"

    iin = f"{dob}{centry_and_gender_code}{individual_number}"

    # Calculate control digit
    checksum_1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
    checksum_2 = [3, 4, 5, 6, 7, 8, 9, 10, 11, 1, 2]
    checksum_1_result = 0
    checksum_2_result = 0
    for i in range(11):
        checksum_1_result += int(iin[i]) * checksum_1[i]
        checksum_2_result += int(iin[i]) * checksum_2[i]
    checksum_1_mod = checksum_1_result % 11
    checksum_2_mod = checksum_2_result % 11
    control_digit = checksum_1_mod
    if checksum_1_mod == 10:
        if checksum_2_mod % 11 == 10:
            # IIN is not active and usable, regenerate IIN
            return generate_iin()
        if checksum_2_mod in range(0, 10):
            control_digit = checksum_2_mod

    return f"{iin}{control_digit}"



kz_iin_validator/iin_validator.py
import datetime as dt
from dataclasses import dataclass
from enum import Enum, auto
from typing import Union, Optional, Tuple

from .exceptions import IINValidateError
from .utils import is_digit_string


@dataclass
class BornDate:
    day: int
    month: int
    year: int
    datetime: dt.datetime


class IINGender(Enum):
    male = auto()
    female = auto()


@dataclass
class IIN:
    iin: str
    is_person: bool
    gender: IINGender
    born_date: BornDate


class ValidatedIIN(IIN):
    is_validated: bool = True


def safe_validate_iin(iin: Union[str, IIN]) -> Tuple[Optional[ValidatedIIN], Exception]:
    """
    Golang-like function for Individual Identification Number (IIN) validation.

    Args:
        iin (Union[str, IIN]): The IIN to be validated. It can be either a string or an IIN object.

    Returns:
        A tuple of `(ValidatedIIN, Exception)`. The first element is `None` if the IIN is not valid, and the second element is an exception if there was an error validating the IIN.

    """
    try:
        result = validate_iin(iin)
    except Exception as ex:
        return None, ex
    else:
        return result, None


# TODO: refactor into separate functions
def validate_iin(iin: Union[str, IIN]) -> ValidatedIIN:
    """
    Validates an Individual Identification Number (IIN).

    Args:
        iin (Union[str, IIN]): The IIN to be validated. It can be either a string or an IIN object.

    Returns:
        ValidatedIIN: The validated IIN as a ValidatedIIN object.

    Raises:
        ValueError: If the provided IIN is invalid.

    """

    if isinstance(iin, IIN):
        iin = iin.iin
    elif not isinstance(iin, str):
        raise IINValidateError(f"Parametr 'iin' must be string, not {type(iin).__name__}")

    if not is_digit_string(iin):
        raise IINValidateError("IIN must contains only numbers")

    if len(iin) != 12:
        raise IINValidateError("IIN must be 12 length")

    # iin helper functions
    def iin_int(index):
        return int(iin[index])

    def iin_int_range(x, y):
        return int(iin[x:y])

    is_person = (iin_int(6) != 0)
    if is_person:
        if iin_int(6) % 2 == 1:
            gender = IINGender.male
        else:
            gender = IINGender.female
    else:
        raise IINValidateError("Can't parse gender information")

    centry_born_code = iin_int(6)
    centry_born_codes_list = {1: 18, 2: 18, 3: 19, 4: 19, 5: 20, 6: 20}
    centry_born = centry_born_codes_list.get(centry_born_code)
    if centry_born is None:
        raise IINValidateError("Can't parse centry from IIN: unknown centry information")

    year = int(f"{centry_born}{iin[0:2]}")
    month = iin_int_range(2, 4)
    day = iin_int_range(4, 6)

    if month not in range(1, 13):
        raise IINValidateError("Month is not valid")

    month_dict = {4: 30, 6: 30, 9: 30, 11: 30, 2: 28}
    day_bound = month_dict.get(month, 31)

    if day_bound == 28 and ((year % 4 == 0 and year % 100 != 0) or year % 400 == 0):
        day_bound = 29

    if day not in range(1, day_bound + 1):
        raise IINValidateError("Invalid day")

    date_string = f"{year}{month}{day}"
    date_format = dt.datetime.strptime(date_string, "%Y%m%d")

    born_date = BornDate(date_format.day, date_format.month, date_format.year, date_format)

    checksum_1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
    checksum_2 = [3, 4, 5, 6, 7, 8, 9, 10, 11, 1, 2]

    checksum_1_result = 0
    checksum_2_result = 0
    for i in range(11):
        checksum_1_result += iin_int(i) * checksum_1[i]
        checksum_2_result += iin_int(i) * checksum_2[i]

    checksum_1_mod = checksum_1_result % 11
    checksum_2_mod = checksum_2_result % 11

    control_digit = checksum_1_mod

    if checksum_1_mod == 10:
        if checksum_2_mod % 11 == 10:
            raise IINValidateError("This IIN is not active and usable!")

        if checksum_2_mod in range(0, 10):
            control_digit = checksum_2_mod

    if control_digit != iin_int(11):
        raise IINValidateError("Not correct control checksum")

    return ValidatedIIN(iin=iin, is_person=is_person, gender=gender, born_date=born_date)
